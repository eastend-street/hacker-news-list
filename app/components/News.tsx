'use client';
import React, { useEffect, useRef } from 'react';
import { Story } from '../../types/story';

export default function News({
  story,
  index,
  isLastStory,
  fetchNextStories,
}: {
  story: Story;
  index: number;
  isLastStory: boolean;
  fetchNextStories: () => void;
}) {
  const newsRef = useRef<HTMLAnchorElement>(null);

  useEffect(() => {
    if (!newsRef?.current || !isLastStory) return;
    const observer = new IntersectionObserver(([entry]) => {
      if (entry.isIntersecting) {
        fetchNextStories();
        observer.unobserve(entry.target);
      }
    });
    observer.observe(newsRef.current);
  }, [isLastStory, fetchNextStories]);

  return story.url ? (
    <a
      href={story.url}
      target="_blank"
      rel="noreferrer"
      ref={newsRef}
      className="hover:underline"
    >
      {index + 1}. {story.title}
    </a>
  ) : (
    <>
      {index + 1}. {story.title}
    </>
  );
}
