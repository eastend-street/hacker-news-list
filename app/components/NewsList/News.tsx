'use client';
import React, { useEffect, useRef } from 'react';
import { Story } from '../../types/story';

export default function News({
  story,
  index,
  isLastStory,
  fetchNextStories,
}: {
  story: Story;
  index: number;
  isLastStory: boolean;
  fetchNextStories: () => void;
}) {
  const newsRef = useRef(null);

  useEffect(() => {
    if (!newsRef?.current || !isLastStory) return;
    const observer = new IntersectionObserver(([entry]) => {
      if (entry.isIntersecting) {
        fetchNextStories();
        observer.unobserve(entry.target);
      }
    });
    observer.observe(newsRef.current);
  }, [isLastStory]);

  return (
    <a
      href={story?.url || `https://news.ycombinator.com/item?id=${story.id}`}
      target="_blank"
      rel="noreferrer"
      ref={newsRef}
      className="hover:underline"
    >
      {index + 1}. {story.title}
    </a>
  );
}
