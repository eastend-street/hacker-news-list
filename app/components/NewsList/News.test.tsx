import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/react';

import * as stories from './News.stories';

const { Default, WithoutUrl, LastStory } = composeStories(stories);

describe('components/News', () => {
  test('should render News component', () => {
    const { container } = render(<Default />);
    const newsAnchorElement = screen.getByRole('link', {
      name: /Forking Chrome to render in a terminal/i,
    });
    expect(newsAnchorElement).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('should have index number with title', () => {
    render(<Default />);
    const newsAnchorElement = screen.getByRole('link', {
      name: /1. Forking Chrome to render in a terminal/i,
    });
    expect(newsAnchorElement).toBeInTheDocument();
  });

  test('should have the news URL as href', () => {
    render(<Default />);
    const newsAnchorElement = screen.getByRole('link', {
      name: /Forking Chrome to render in a terminal/i,
    });
    expect(newsAnchorElement).toHaveAttribute(
      'href',
      'https://fathy.fr/carbonyl'
    );
  });

  test('should have the news.ycombinator.com as href when the story does not have an URL', () => {
    render(<WithoutUrl />);
    const newsAnchorElement = screen.getByRole('link', {
      name: /Ask HN: Why isn't radiation used to sterilize raw foods (like sashimi, eggs)?/i,
    });
    expect(newsAnchorElement).toHaveAttribute(
      'href',
      'https://news.ycombinator.com/item?id=34573675'
    );
  });
});
