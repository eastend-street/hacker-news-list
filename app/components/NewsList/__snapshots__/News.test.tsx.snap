// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/News should render News component 1`] = `
<div>
  <a
    class="hover:underline"
    href="https://fathy.fr/carbonyl"
    rel="noreferrer"
    target="_blank"
  >
    1
    . 
    Forking Chrome to render in a terminal
  </a>
</div>
`;
